### PR TITLE
Merge 2.12.x to 2.13.x [ci: last-only]

### PR DESCRIPTION
**Pull request by retronym**
Also includes a commit to cleanup some temporary duplication in HashMap (a rebased version of @paplorinc's #5567)

```
% export MB=$(git merge-base origin/2.12.x origin/2.13.x)

% git log --graph --oneline --decorate $MB..origin/2.12.x | gist --filename git.log
https://gist.github.com/d661e319e6d66c865c385d4560db12a6
% git co -b merge/2.12.x-to-2.13.x-20161130
Switched to a new branch 'merge/2.12.x-to-2.13.x-20161130'

% git merge origin/2.12.x
Updating 10c609e..0339663
Fast-forward
 META-INF/MANIFEST.MF                                                               |  51 --------
 bincompat-backward.whitelist.conf                                                  | 213 +-----------------------------
 bincompat-forward.whitelist.conf                                                   | 478 ++-----------------------------------------------------------------
 build.sbt                                                                          | 159 ++++++++++++----------
 project/MiMa.scala                                                                 |   8 +-
 project/Osgi.scala                                                                 |  19 ++-
 project/PartestUtil.scala                                                          |   5 +-
 scripts/jobs/integrate/bootstrap                                                   |   2 +-
 spec/01-lexical-syntax.md                                                          |   2 +-
 spec/03-types.md                                                                   |  10 +-
 spec/04-basic-declarations-and-definitions.md                                      |   2 +-
 spec/05-classes-and-objects.md                                                     |   6 +-
 spec/06-expressions.md                                                             |  14 +-
 spec/11-annotations.md                                                             |   8 +-
 spec/12-the-scala-standard-library.md                                              |   8 +-
 spec/13-syntax-summary.md                                                          |   2 +-
 spec/README.md                                                                     |   4 +-
 spec/_layouts/default.yml                                                          |   4 +-
 src/build/genprod.scala                                                            |   2 +-
 src/compiler/rootdoc.txt                                                           |   7 +-
 src/compiler/scala/tools/nsc/Global.scala                                          |   5 +-
 src/compiler/scala/tools/nsc/ast/parser/Scanners.scala                             |  13 +-
 src/compiler/scala/tools/nsc/backend/jvm/BCodeBodyBuilder.scala                    |  13 +-
 src/compiler/scala/tools/nsc/backend/jvm/BCodeSkelBuilder.scala                    |   1 -
 src/compiler/scala/tools/nsc/backend/jvm/BCodeSyncAndTry.scala                     |  48 +++++--
 src/compiler/scala/tools/nsc/backend/jvm/BTypes.scala                              |  13 +-
 src/compiler/scala/tools/nsc/backend/jvm/BTypesFromSymbols.scala                   |   2 +-
 src/compiler/scala/tools/nsc/backend/jvm/BackendReporting.scala                    |  76 ++++++-----
 src/compiler/scala/tools/nsc/backend/jvm/opt/CallGraph.scala                       |  42 +++---
 src/compiler/scala/tools/nsc/backend/jvm/opt/ClosureOptimizer.scala                |   4 +-
 src/compiler/scala/tools/nsc/backend/jvm/opt/InlineInfoAttribute.scala             |   2 +-
 src/compiler/scala/tools/nsc/backend/jvm/opt/Inliner.scala                         | 311 ++++++++++++++++++++++++++++++-------------
 src/compiler/scala/tools/nsc/backend/jvm/opt/InlinerHeuristics.scala               | 209 +++++++++++++++++------------
 src/compiler/scala/tools/nsc/classpath/VirtualDirectoryClassPath.scala             |  12 +-
 src/compiler/scala/tools/nsc/symtab/classfile/ClassfileParser.scala                |   8 +-
 src/compiler/scala/tools/nsc/transform/AccessorSynthesis.scala                     |   2 +-
 src/compiler/scala/tools/nsc/transform/Erasure.scala                               |   1 -
 src/compiler/scala/tools/nsc/transform/Fields.scala                                |  29 +++--
 src/compiler/scala/tools/nsc/transform/Mixin.scala                                 |   2 +-
 src/compiler/scala/tools/nsc/transform/SpecializeTypes.scala                       |   3 +-
 src/compiler/scala/tools/nsc/transform/UnCurry.scala                               |  95 +++++---------
 src/compiler/scala/tools/nsc/transform/patmat/Logic.scala                          |   2 +-
 src/compiler/scala/tools/nsc/transform/patmat/MatchAnalysis.scala                  |   2 +-
 src/compiler/scala/tools/nsc/typechecker/Contexts.scala                            |   2 +-
 src/compiler/scala/tools/nsc/typechecker/MethodSynthesis.scala                     |   4 +
 src/compiler/scala/tools/nsc/typechecker/Namers.scala                              |   2 +-
 src/compiler/scala/tools/nsc/typechecker/PatternTypers.scala                       |   2 +-
 src/compiler/scala/tools/nsc/typechecker/RefChecks.scala                           |   4 +-
 src/compiler/scala/tools/nsc/typechecker/Typers.scala                              |  35 ++++-
 src/interactive/scala/tools/nsc/interactive/Global.scala                           |   3 +-
 src/library/scala/Array.scala                                                      |   3 +-
 src/library/scala/Immutable.scala                                                  |   2 +-
 src/library/scala/Int.scala                                                        |   4 +-
 src/library/scala/Option.scala                                                     |   2 +-
 src/library/scala/Product1.scala                                                   |   2 +-
 src/library/scala/Product10.scala                                                  |   2 +-
 src/library/scala/Product11.scala                                                  |   2 +-
 src/library/scala/Product12.scala                                                  |   2 +-
 src/library/scala/Product13.scala                                                  |   2 +-
 src/library/scala/Product14.scala                                                  |   2 +-
 src/library/scala/Product15.scala                                                  |   2 +-
 src/library/scala/Product16.scala                                                  |   2 +-
 src/library/scala/Product17.scala                                                  |   2 +-
 src/library/scala/Product18.scala                                                  |   2 +-
 src/library/scala/Product19.scala                                                  |   2 +-
 src/library/scala/Product2.scala                                                   |   2 +-
 src/library/scala/Product20.scala                                                  |   2 +-
 src/library/scala/Product21.scala                                                  |   2 +-
 src/library/scala/Product22.scala                                                  |   2 +-
 src/library/scala/Product3.scala                                                   |   2 +-
 src/library/scala/Product4.scala                                                   |   2 +-
 src/library/scala/Product5.scala                                                   |   2 +-
 src/library/scala/Product6.scala                                                   |   2 +-
 src/library/scala/Product7.scala                                                   |   2 +-
 src/library/scala/Product8.scala                                                   |   2 +-
 src/library/scala/Product9.scala                                                   |   2 +-
 src/library/scala/collection/Iterator.scala                                        |  27 ++--
 src/library/scala/collection/Parallelizable.scala                                  |   2 +-
 src/library/scala/collection/SetLike.scala                                         |   4 +-
 src/library/scala/collection/SortedSet.scala                                       |   2 +-
 src/library/scala/collection/generic/GenTraversableFactory.scala                   |   2 +-
 src/library/scala/collection/immutable/List.scala                                  |   4 +-
 src/library/scala/collection/immutable/SetProxy.scala                              |   3 +-
 src/library/scala/collection/immutable/SortedSet.scala                             |   2 +-
 src/library/scala/collection/immutable/Vector.scala                                |   9 +-
 src/library/scala/collection/mutable/ArrayBuffer.scala                             |   2 +-
 src/library/scala/collection/mutable/HashMap.scala                                 |  31 +++++
 src/library/scala/collection/mutable/HashTable.scala                               |  85 ++++--------
 src/library/scala/collection/mutable/History.scala                                 |   2 +-
 src/library/scala/collection/mutable/PriorityQueue.scala                           |   2 +-
 src/library/scala/collection/mutable/ResizableArray.scala                          |   2 +-
 src/library/scala/collection/mutable/ReusableBuilder.scala                         |   4 +-
 src/library/scala/collection/parallel/TaskSupport.scala                            |   4 +-
 src/library/scala/collection/parallel/immutable/ParRange.scala                     |   1 +
 src/library/scala/concurrent/SyncVar.scala                                         |   4 +-
 src/library/scala/concurrent/duration/Duration.scala                               |  14 +-
 src/library/scala/concurrent/impl/ExecutionContextImpl.scala                       |   2 +-
 src/library/scala/concurrent/impl/Promise.scala                                    |   2 +-
 src/library/scala/inline.scala                                                     |   2 +-
 src/library/scala/io/Source.scala                                                  |   2 +-
 src/library/scala/math/BigInt.scala                                                |   4 +-
 src/library/scala/noinline.scala                                                   |   2 +-
 src/library/scala/reflect/ClassManifestDeprecatedApis.scala                        |   4 +-
 src/library/scala/sys/process/ProcessBuilder.scala                                 |   2 +-
 src/library/scala/sys/process/ProcessBuilderImpl.scala                             |  14 +-
 src/library/scala/sys/process/ProcessImpl.scala                                    |  34 ++---
 src/library/scala/sys/process/package.scala                                        |   4 +-
 src/library/scala/util/Either.scala                                                |   2 +-
 src/library/scala/util/Properties.scala                                            |   6 -
 src/manual/scala/man1/scalac.scala                                                 |   4 +-
 src/manual/scala/man1/scalap.scala                                                 |   2 +-
 src/partest-extras/scala/tools/partest/JavapTest.scala                             |   2 +-
 src/partest-extras/scala/tools/partest/ReplTest.scala                              |   2 +-
 src/reflect/scala/reflect/api/Internals.scala                                      |   2 +-
 src/reflect/scala/reflect/api/StandardDefinitions.scala                            |   2 +-
 src/reflect/scala/reflect/api/Types.scala                                          |   2 +-
 src/reflect/scala/reflect/internal/AnnotationInfos.scala                           |   2 +-
 src/reflect/scala/reflect/internal/BaseTypeSeqs.scala                              |   2 +-
 src/reflect/scala/reflect/internal/Constants.scala                                 |   4 +-
 src/reflect/scala/reflect/internal/Names.scala                                     |   4 +-
 src/reflect/scala/reflect/internal/StdAttachments.scala                            |   3 +
 src/reflect/scala/reflect/internal/Symbols.scala                                   |  14 +-
 src/reflect/scala/reflect/internal/TreeGen.scala                                   |   2 +-
 src/reflect/scala/reflect/internal/TreeInfo.scala                                  |   3 +-
 src/reflect/scala/reflect/internal/Types.scala                                     |   5 +-
 src/reflect/scala/reflect/internal/pickling/UnPickler.scala                        |  43 ++----
 src/reflect/scala/reflect/internal/tpe/FindMembers.scala                           |   4 +-
 src/reflect/scala/reflect/internal/tpe/TypeMaps.scala                              |   4 +-
 src/reflect/scala/reflect/internal/transform/UnCurry.scala                         |  69 +++++++++-
 src/reflect/scala/reflect/internal/util/AbstractFileClassLoader.scala              |  28 ++--
 src/reflect/scala/reflect/internal/util/SourceFile.scala                           |  23 ++--
 src/reflect/scala/reflect/runtime/JavaUniverse.scala                               |   2 +-
 src/reflect/scala/reflect/runtime/JavaUniverseForce.scala                          |   2 +
 src/reflect/scala/reflect/runtime/SynchronizedSymbols.scala                        |   2 +
 src/repl/scala/tools/nsc/interpreter/IMain.scala                                   |   2 +-
 src/repl/scala/tools/nsc/interpreter/InteractiveReader.scala                       |   2 +-
 src/repl/scala/tools/nsc/interpreter/Phased.scala                                  |   2 +-
 src/repl/scala/tools/nsc/interpreter/ReplGlobal.scala                              |  13 ++
 src/scaladoc/scala/tools/nsc/doc/base/MemberLookupBase.scala                       |   2 +-
 src/scaladoc/scala/tools/nsc/doc/html/page/diagram/DotDiagramGenerator.scala       |   2 +-
 src/scaladoc/scala/tools/nsc/doc/html/resource/lib/scheduler.js                    |   4 +-
 src/scaladoc/scala/tools/nsc/doc/html/resource/lib/template.css                    |   2 +-
 src/scaladoc/scala/tools/nsc/doc/model/Entity.scala                                |   8 +-
 src/scaladoc/scala/tools/nsc/doc/model/ModelFactory.scala                          |   4 +-
 test/benchmarks/README.md                                                          |   7 +-
 test/benchmarks/build.sbt                                                          |   4 +-
 test/benchmarks/project/plugins.sbt                                                |   2 +-
 test/benchmarks/src/main/scala/scala/BitManipulationBenchmark.scala                | 170 ++++++++++++++++++++++++
 test/benchmarks/src/main/scala/scala/collection/immutable/VectorMapBenchmark.scala |  32 +++++
 test/benchmarks/src/main/scala/scala/collection/mutable/HashMapBenchmark.scala     |  70 ++++++++++
 test/files/jvm/future-spec/FutureTests.scala                                       |   2 +-
 test/files/jvm/serialization-new.check                                             |   8 +-
 test/files/jvm/serialization.check                                                 |   8 +-
 test/files/jvm/varargs-separate-bytecode.check                                     |   1 +
 test/files/jvm/varargs-separate-bytecode/AbstractProps_1.scala                     |   8 ++
 test/files/jvm/varargs-separate-bytecode/Props_2.scala                             |   3 +
 test/files/jvm/varargs-separate-bytecode/Test.scala                                |  15 +++
 test/files/neg/sabin2.check                                                        |   2 +-
 test/files/neg/sealed-final-neg.check                                              |   6 +-
 test/files/neg/t0764.check                                                         |   2 +-
 test/files/neg/t1010.check                                                         |   2 +-
 test/files/neg/t5120.check                                                         |   2 +-
 test/files/neg/t6829.check                                                         |  14 +-
 test/files/neg/t9953.check                                                         |   6 +
 test/files/neg/t9953.flags                                                         |   1 +
 test/files/neg/t9953.scala                                                         |  13 ++
 test/files/pos/sd268.scala                                                         |  17 +++
 test/files/pos/t10009.scala                                                        |   6 +
 test/files/pos/t6734.scala                                                         |  17 +++
 test/files/pos/t6978.flags                                                         |   1 +
 test/files/pos/t6978/J.java                                                        |   5 +
 test/files/pos/t6978/S.scala                                                       |   7 +
 test/files/run/reflection-mem-typecheck.scala                                      |  28 ----
 test/files/run/repl-inline.check                                                   |  11 ++
 test/files/run/repl-inline.scala                                                   |  27 ++++
 test/files/run/sd275-java/A.java                                                   |   5 +
 test/files/run/sd275-java/DeleteMe.java                                            |   4 +
 test/files/run/sd275-java/LeaveMe.java                                             |   3 +
 test/files/run/sd275-java/Test.scala                                               |  39 ++++++
 test/files/run/sd275.scala                                                         |  60 +++++++++
 test/files/run/t10009.scala                                                        |  28 ++++
 test/files/run/t10032.check                                                        |  82 ++++++++++++
 test/files/run/t10032.scala                                                        | 164 +++++++++++++++++++++++
 test/files/run/t10059/A.java                                                       |   3 +
 test/files/run/t10059/Test.scala                                                   |   9 ++
 test/files/run/t5125b.check                                                        |   3 +
 test/files/run/t5125b.scala                                                        |  12 ++
 test/files/run/t7747-repl.check                                                    |   6 +-
 test/files/run/t8433.check                                                         |   2 +
 test/files/run/t8433.scala                                                         |  46 +++++++
 test/files/run/t9375.scala                                                         |   2 +-
 test/files/run/t9689.check                                                         |  14 ++
 test/files/run/t9689/Test_2.scala                                                  |  12 ++
 test/files/run/t9689/bug_1.scala                                                   |   8 ++
 test/files/run/t9814.scala                                                         |  28 ++++
 test/files/run/t9915/C_1.java                                                      |  20 +++
 test/files/run/t9915/Test_2.scala                                                  |  12 ++
 test/files/run/t9944.check                                                         |  12 ++
 test/files/run/t9944.scala                                                         |   7 +
 test/junit/scala/collection/IteratorTest.scala                                     |   6 +
 test/junit/scala/collection/parallel/immutable/ParRangeTest.scala                  |  15 +++
 test/junit/scala/reflect/internal/util/SourceFileTest.scala                        |  19 +++
 test/junit/scala/sys/process/{t7350.scala => PipedProcessTest.scala}               |   1 +
 test/junit/scala/sys/process/ProcessTest.scala                                     |  25 ++++
 test/junit/scala/tools/nsc/backend/jvm/opt/CallGraphTest.scala                     |  13 +-
 test/junit/scala/tools/nsc/backend/jvm/opt/InlineWarningTest.scala                 |   6 +-
 test/junit/scala/tools/nsc/backend/jvm/opt/InlinerIllegalAccessTest.scala          |  22 ++--
 test/junit/scala/tools/nsc/backend/jvm/opt/InlinerSeparateCompilationTest.scala    |   2 +-
 test/junit/scala/tools/nsc/backend/jvm/opt/InlinerTest.scala                       | 159 ++++++++++++++++++++--
 test/junit/scala/tools/nsc/classpath/VirtualDirectoryClassPathTest.scala           |  41 ++++++
 test/junit/scala/tools/nsc/interpreter/CompletionTest.scala                        |  21 ++-
 test/junit/scala/util/SpecVersionTest.scala                                        |  11 +-
 212 files changed, 2404 insertions(+), 1580 deletions(-)
 delete mode 100644 META-INF/MANIFEST.MF
 create mode 100644 test/benchmarks/src/main/scala/scala/BitManipulationBenchmark.scala
 create mode 100644 test/benchmarks/src/main/scala/scala/collection/immutable/VectorMapBenchmark.scala
 create mode 100644 test/benchmarks/src/main/scala/scala/collection/mutable/HashMapBenchmark.scala
 create mode 100644 test/files/jvm/varargs-separate-bytecode.check
 create mode 100644 test/files/jvm/varargs-separate-bytecode/AbstractProps_1.scala
 create mode 100644 test/files/jvm/varargs-separate-bytecode/Props_2.scala
 create mode 100644 test/files/jvm/varargs-separate-bytecode/Test.scala
 create mode 100644 test/files/neg/t9953.check
 create mode 100644 test/files/neg/t9953.flags
 create mode 100644 test/files/neg/t9953.scala
 create mode 100644 test/files/pos/sd268.scala
 create mode 100644 test/files/pos/t10009.scala
 create mode 100644 test/files/pos/t6734.scala
 create mode 100644 test/files/pos/t6978.flags
 create mode 100644 test/files/pos/t6978/J.java
 create mode 100644 test/files/pos/t6978/S.scala
 delete mode 100644 test/files/run/reflection-mem-typecheck.scala
 create mode 100644 test/files/run/repl-inline.check
 create mode 100644 test/files/run/repl-inline.scala
 create mode 100644 test/files/run/sd275-java/A.java
 create mode 100644 test/files/run/sd275-java/DeleteMe.java
 create mode 100644 test/files/run/sd275-java/LeaveMe.java
 create mode 100644 test/files/run/sd275-java/Test.scala
 create mode 100644 test/files/run/sd275.scala
 create mode 100644 test/files/run/t10009.scala
 create mode 100644 test/files/run/t10032.check
 create mode 100644 test/files/run/t10032.scala
 create mode 100644 test/files/run/t10059/A.java
 create mode 100644 test/files/run/t10059/Test.scala
 create mode 100644 test/files/run/t8433.check
 create mode 100644 test/files/run/t8433.scala
 create mode 100644 test/files/run/t9689.check
 create mode 100644 test/files/run/t9689/Test_2.scala
 create mode 100644 test/files/run/t9689/bug_1.scala
 create mode 100644 test/files/run/t9814.scala
 create mode 100644 test/files/run/t9915/C_1.java
 create mode 100644 test/files/run/t9915/Test_2.scala
 create mode 100755 test/files/run/t9944.check
 create mode 100644 test/files/run/t9944.scala
 create mode 100644 test/junit/scala/collection/parallel/immutable/ParRangeTest.scala
 rename test/junit/scala/sys/process/{t7350.scala => PipedProcessTest.scala} (99%)
 create mode 100644 test/junit/scala/sys/process/ProcessTest.scala
 create mode 100644 test/junit/scala/tools/nsc/classpath/VirtualDirectoryClassPathTest.scala
 ```

 
 for containing ellipses (\ldots).
 elements, such as identifiers that are italicized or have prime
 e′′), ellipses (\ldots), etc.

### CODE REVIEW

**Comment by retronym** [comment]
Review by @SethTisue

**Comment by SethTisue** [comment]
LGTM, but Scabot seems hosed

**Comment by SethTisue** [comment]
ticket on whatever's up with Scabot here: https://github.com/scala/scala-dev/issues/279

File: src/library/scala/collection/mutable/HashTable.scala [hunk 1]
```diff
@@ -131,7 +131,7 @@ trait HashTable[A, Entry >: Null <: HashEntry[A, Entry]] extends HashTable.HashU
   protected def findEntry(key: A): Entry =
     findEntry0(key, index(elemHashCode(key)))
 
-  private[this] def findEntry0(key: A, h: Int): Entry = {
+  protected[this] final def findEntry0(key: A, h: Int): Entry = {
```
Comment (l0rinc) [inline src/library/scala/collection/mutable/HashTable.scala:134]: could this be causing the build failure?

**Comment by adriaanm** [comment]
May be because I didn't implement paging for the github API and this PR has a lot of commits (so it can't get to the last one?)

**Comment by SethTisue** [comment]
that's an excellent guess. now that we have a theory, personally I'd be content to just `/nothingtoseehere` it and if there's some regression let the 2.13 nightly build catch it

**Comment by adriaanm** [comment]
Agreed.

**Comment by SethTisue** [comment]
perhaps the cause for the failures at #5581 / #5582 is somewhere in here

**Pull request merged**